### PR TITLE
Update build/sdk to net8.0

### PR DESCRIPTION
**Pull request by rhysparry**
# Background

To help support migrating to .NET 8 this change updates the SDK to use net8.0 when building and for the Nuke build project.

This does not change the target framework of Tentacle, which will continue to target the existing frameworks in this PR.

# Results

- Nuke Build updated from 6.0.3 -> 8.0.0
- OctoVersion updated from 0.2.1058 -> 0.3.400
- OctopusTools updated from 7.4.3289 -> 9.1.7
- Fixed/suppressed warnings for issues raised by new compiler tooling (primarily defining the actual nullability of variables)

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [x] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/main/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [x] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [x] I have considered appropriate testing for my change.

### CODE REVIEW

File: build/Build.Pack.cs [hunk 1]
```diff
@@ -520,7 +521,16 @@ string ConstructRedHatPackageFilename(string packageName, string architecture)
...
+
+            OctopusTasks.OctopusPack(o => o
+                .SetId("Octopus.Tentacle.CrossPlatformBundle")
+                .SetVersion(FullSemVer)
+                .SetBasePath(workingDirectory)
+                .SetOutputFolder(ArtifactsDirectory / "nuget")
+                .SetAuthors(author)
+                .SetTitle(title)
+                .SetDescription(description)
+            );
```
Comment (rhysparry) [inline build/Build.Pack.cs:533]: The string interpolation led to issues with extraneous quotes. Using the built-in Nuke build support is more reliable

File: build/Build.cs [hunk 2]
```diff
@@ -20,7 +20,6 @@
 using static Nuke.Common.IO.FileSystemTasks;
 using static Nuke.Common.Tools.DotNet.DotNetTasks;
 
-[CheckBuildProjectConfigurations]
```
Comment (rhysparry) [inline build/Build.cs:23]: This is no longer required.

File: build/Build.cs [hunk 3]
```diff
@@ -51,33 +50,25 @@ public Build()
...
     readonly OctoVersionInfo OctoVersionInfo = null!;
 
     [Parameter] string TestFramework = "";
     [Parameter] string TestRuntime = "";
     [Parameter] string TestFilter = "";
 
-    [PackageExecutable(
-        packageId: "azuresigntool",
-        packageExecutable: "azuresigntool.dll")]
-    public static Tool AzureSignTool = null!;
```
Comment (rhysparry) [inline build/Build.cs:69]: We're using the built-in support for AzureSignTool now.

File: build/Build.cs [hunk 4]
```diff
@@ -51,33 +50,25 @@ public Build()
...
-    [PackageExecutable(
+    [NuGetPackage(
         packageId: "wix",
         packageExecutable: "heat.exe")]
     readonly Tool WiXHeatTool = null!;
 
-    [PackageExecutable(
-        packageId: "OctopusTools",
-        packageExecutable: "octo.exe")]
-    readonly Tool OctoCliTool = null!;
```
Comment (rhysparry) [inline build/Build.cs:79]: We're using the built-in support for the Octo CLI

File: build/Signing.cs [hunk 5]
```diff
@@ -106,19 +107,17 @@ static void SignWithAzureSignTool(AbsolutePath[] files)
...
+                    AzureSignToolTasks.AzureSignTool(a => a
+                        .SetKeyVaultUrl(Build.AzureKeyVaultUrl)
+                        .SetKeyVaultClientId(Build.AzureKeyVaultAppId)
+                        .SetKeyVaultTenantId(Build.AzureKeyVaultTenantId)
+                        .SetKeyVaultClientSecret(Build.AzureKeyVaultAppSecret)
+                        .SetKeyVaultCertificateName(Build.AzureKeyVaultCertificateName)
+                        .SetFileDigest(AzureSignToolDigestAlgorithm.sha256)
+                        .SetTimestampRfc3161Url(timestampUrl)
+                        .SetFiles(files.Select(f => f.ToString()))
+                    );
```
Comment (rhysparry) [inline build/Signing.cs:119]: Using the built-in support avoids issues with double-escaping of quotes.

File: build/_build.csproj [hunk 6]
```diff
@@ -11,16 +11,16 @@
...
 
   <ItemGroup>
-    <PackageDownload Include="AzureSignTool" Version="[3.0.0]" />
+    <PackageDownload Include="AzureSignTool" Version="[5.0.0]" />
     <PackageDownload Include="Chocolatey" Version="[0.10.14]" />
     <PackageDownload Include="WiX" Version="[3.11.2]" />
-    <PackageDownload Include="OctopusTools" Version="[7.4.3289]" />
-    <PackageDownload Include="OctoVersion.Tool" Version="[0.2.1058]" />
+    <PackageDownload Include="Octopus.DotNet.Cli" Version="[9.1.7]" />
+    <PackageDownload Include="Octopus.OctoVersion.Tool" Version="[0.3.400]" />
```
Comment (rhysparry) [inline build/_build.csproj:23]: These are the preferred packages (the previous ones are deprecated)

File: build.ps1 [hunk 7]
```diff
@@ -28,22 +28,30 @@ $env:DOTNET_MULTILEVEL_LOOKUP = 0
...
+# Avoid issues where LASTEXITCODE does not appear to be set
+$LASTEXITCODE = 0
+# End Octopus Modification
+
 function ExecSafe([scriptblock] $cmd) {
     & $cmd
     if ($LASTEXITCODE) { exit $LASTEXITCODE }
 }
 
 # If dotnet CLI is installed globally and it matches requested version, use for execution
```
Comment (LukeButters) [inline build.ps1:41]: A comment here that on TC we always get our own dotnet.

File: build.ps1 [hunk 8]
```diff
@@ -60,6 +68,12 @@ else {
     } else {
         ExecSafe { & $DotNetInstallFile -InstallDir $DotNetDirectory -Version $DotNetVersion -NoPath }
     }
+
+    # Octopus Modification
+    # Install the .NET 6.0 runtime as well
```
Comment (LukeButters) [inline build.ps1:73]: Is this also for tests? Can it have a comment?

File: build/Build.Pack.cs [hunk 9]
```diff
@@ -30,9 +31,9 @@ partial class Build
     [Parameter("Specifies the platforms to build the docker images in. Multiple platforms must be comma-separated. Defaults to 'linux/arm64,linux/amd64'.",
         Name = "DockerPlatform")]
     string DockerPlatform = "linux/arm64,linux/amd64";
-    
-    [PathExecutable]
-    readonly Tool Multipass;
+
+    [PathVariable]
```
Comment (LukeButters) [inline build/Build.Pack.cs:35]: Why this change?

File: build/Build.Pack.cs [hunk 10]
```diff
@@ -122,13 +123,13 @@ void CreateLinuxPackages(string runtimeId)
 
                 var debOutputDirectory = BuildDirectory / "deb" / runtimeId / "output";
 
-                FileSystemTasks.EnsureExistingDirectory(ArtifactsDirectory / "deb");
+                (ArtifactsDirectory / "deb").CreateDirectory();
```
Comment (LukeButters) [inline build/Build.Pack.cs:126]: The extension method is nice.

File: build/Build.Pack.cs [hunk 11]
```diff
@@ -520,7 +521,16 @@ string ConstructRedHatPackageFilename(string packageName, string architecture)
             const string description = "The deployment agent that is installed on each machine you plan to deploy to using Octopus.";
             const string author = "Octopus Deploy";
             const string title = "Octopus Tentacle cross platform bundle";
-            OctoCliTool($@"pack --id=Octopus.Tentacle.CrossPlatformBundle --version={FullSemVer} --basePath={workingDirectory} --outFolder={ArtifactsDirectory / "nuget"} --author=""{author}"" --title=""{title}"" --description=""{description}""");
+
+            OctopusTasks.OctopusPack(o => o
```
Comment (LukeButters) [inline build/Build.Pack.cs:525]: This is really nice! How did you find it and know to use it?

File: build/Build.cs [hunk 12]
```diff
@@ -325,7 +316,7 @@ void RunBuildFor(string framework, string runtimeId)
             .SetProject(SourceDirectory / "Tentacle.sln")
             .SetConfiguration(configuration)
             .SetFramework(framework)
-            //.SetSelfContained(true)
+            .SetSelfContained(true)
```
Comment (LukeButters) [inline build/Build.cs:319]: Why did this end up changing?

File: build/test-scripts/integration/common/setup-vm-agent.sh [hunk 13]
```diff
@@ -5,4 +5,4 @@
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
 
 source $SCRIPT_DIR/dotnet-install.sh
-install_dotnet 6.0
+install_dotnet --runtime 6.0 --sdk 8.0
```
Comment (LukeButters) [inline build/test-scripts/integration/common/setup-vm-agent.sh:8]: Why does the runtime differ from the sdk? oh is it 8  compiling for 6? Or is runtime 6.0 just needed for tests? Whatever the reason some helpful doco will help the next poor developer

File: global.json [hunk 14]
```diff
@@ -1,6 +1,6 @@
 {
   "sdk": {
-    "version": "6.0.300",
+    "version": "8.0.303",
```
Comment (LukeButters) [inline global.json:3]: yay

**Review by LukeButters — APPROVED** [review]

File: build/Build.Pack.cs (hunk 9 above)
Comment (rhysparry) [reply build/Build.Pack.cs:35]: Changes in the Nuke Attribute. `PathVariableAttribute` replaces the obsolete `PathExecutableAttribute`

https://github.com/nuke-build/nuke/blob/6ed9e4d9c7476444eb6b984591d1364b47723144/source/Nuke.Common/Attributes/PathVariableAttribute.cs#L55-L56

File: build/Build.Pack.cs (hunk 11 above)
Comment (rhysparry) [reply build/Build.Pack.cs:525]: While trying to understand what was going on in the argument parsing (it was having issues with quotes in arguments I downloaded the Nuke source and discovered this while browsing.

File: build/test-scripts/integration/common/setup-vm-agent.sh (hunk 13 above)
Comment (rhysparry) [reply build/test-scripts/integration/common/setup-vm-agent.sh:8]: I'll add a comment. Basically this is saying install the 6.0 runtime and the 8.0 sdk. There is some hyper simplistic argument parsing going on.

File: build/Build.cs (hunk 12 above)
Comment (rhysparry) [reply build/Build.cs:319]: This was perhaps a pre-emptive strike at addressing this warning:
> warning NETSDK1201: For projects targeting .NET 8.0 and higher, specifying a RuntimeIdentifier will no longer produce a self contained app by default. To continue building self-contained apps, set the SelfContained property to true or use the --self-contained argument. 

The warning shows up for builds on net6 interestingly...

File: build/Build.Pack.cs (hunk 11 above)
Comment (LukeButters) [reply build/Build.Pack.cs:525]: noice

**Review by APErebus — APPROVED** [review]
LGTM for the Kubernetes integration tests changes as codeowners. I didn't check the rest :)
